keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,228 @@
+
+This proposal is grounded on the principle that a trace context is attached to and propagated with end-user intent.  When the status of an object is updated to its desired state, the end-user's intent for that object has been fulfilled.  Controllers must "end" tracing for an object when it reaches its desired state.  To accomplish this, Controllers must update the trace context of an object when updating the status of an object from an undesired to a desired state.  For objects that report a status that can reach a desired state, this limits traces to just the actions taken by controllers in the fulfillment of the end-user's intent, and prevents traces from spanning an indefinite period of time.
+
+Components should plumb the context through reconciliation functions, rather than storing and looking up trace contexts globally so that each attempt to reconcile desired and actual state uses the context associated with _that_ desired state through the entire attempt.  If multiple components are involved in reconciling a single object, one may act on the new trace context before the other, but each trace is still representative of the work done to reconcile to the corresponding desired state. Given this model, we guarantee that each trace contains the actions taken to reconcile toward a single desired state.
+
+High-level processes, such as starting a pod or restarting a failed container, could be interrupted before completion by an update to the desired state. While this leaves a "partial" trace for the first process, it is the most accurate representation of the work and timing of reconciling desired and actual state.
+
+#### End-User Behavior
+
+A previous iteration of this proposal suggested controllers should export a "Root Span" when ending a trace (described in [Controller Behavior](#controller-behavior) above).  However, that would limit a trace to being associated with a single object, since a "Root Span" defines the scope of the trace.  More generally, we shouldn't assume that the creation or update of a single object represents the entirety of end-user intent.  The user or system using kubernetes determines what the user intent is, not kubernetes controllers.
+
+Tracing in a kubernetes cluster must be a composable component within a larger system, and allow external users or systems to define the "Root Span" that defines and bounds the scope of a trace.
+
+For example, a user could define a Pod Creation trace doing something like the following:
+```golang
+ctx, span := trace.StartSpan(context.Background(), “create-my-pod”)
+defer span.End()
+ctx = traceutil.WithSpanContext(ctx, span)
+pod, err := c.CoreV1().Pod(myPod.Namespace).Create(ctx, myPod)
+if err != nil {
+    return err
+}
+waitForPodToBeRunning(ctx, myPod)
+return nil
+```
+
+### In-tree changes
```

Comment:
I've read through this section and this potentially can touch different pieces of codebase. We should probably discuss with sig-apimachinery for their lessons learnt, but I think it may benefit from feature branch.